### ISSUE TITLE
Add portable version of Atom

### ISSUE THREAD

**Issue by d10** [atom/atom#2939]
I don’t really stick to one computer to program, and some computers don’t have editors. Currently I use Sublime Text portable, but it would be great if there was a portable version of Atom.

**Comment by Syenta** [comment]
Atom is a great little editor but as someone who needs to code whilst on the move eg trains, buses, etc and have it installed on a USB drive I could do with Atom being portable. To fill my needs for the present I use a portable version of Notepad++ but ultimately would prefer just to work in Atom, hopefully this will become a reality soon. By the way thanks for the great work you have done to date.

**Comment by paulcbetts** [comment]
Some folks have committed patches to Squirrel to allow you to use it in a Portable way, if the Atom Team wants to support this scenario I'd help out with it

**Comment by benogle** [comment]
@paulcbetts We would love to support this. We have a lot on our plate already, and honestly none of us are super windows inclined. Any help on this would be much appreciated. :+1:

**Comment by lee-dohm** [comment]
Per the discussion on #4980, we should cover Linux too.

**Comment by tknerr** [comment]
@benogle @paulcbetts that sounds awesome! I would be happy to jump in testing as soon as a preliminary version is available.

**Comment by CumpsD** [comment]
@paulcbetts is this something you would be doing, reading @benogle's comment, or is this something one of us should provide a PR for where you can help with?

Just to make sure we're not waiting for each other :)

**Comment by paulcbetts** [comment]
@CumpsD Lemme look into what it would entail, then I'll ping back this thread

**Comment by stavpup** [comment]
Is it possible to reduce the number of files for the portable version? I count about 8000+ files and that is not very sane for a flash drive. Can it be reduced to a big binary and a couple of files?

**Comment by paulcbetts** [comment]
@stavpup Atom could do this if they switched to using ASAR, but that's a non-trivial undertaking and I don't know if/when they plan to do that.

**Comment by DominikLevitsky** [comment]
Hello. Any updates on this?

**Comment by nuxonom** [comment]
Hey everyone, I've successfully gotten a portable 0.176.0 windows installation working, based on @bclindner's post: https://github.com/atom/atom/issues/2939#issuecomment-62586621
The only issue I've run into so far is that it'll still display the user's home path for the location of the `.atom` folder. Functionally, this will make atom use the exe's folder for `.atom`.

Here's the updated info:
Note: change every instance of `~/.atom` to `./.atom`

`resources\app\apm\.apmrc`
line 1: `cache = ~/.atom/.apm`

`resources\app\src\browser\atom-application.js`
line 663: `configDirPath: fs.absolute('~/.atom'),`

`resources\app\src\atom.js`
line 114: `return this.configDirPath != null ? this.configDirPath : this.configDirPath = fs.absolute('~/.atom');`

`resources\app\src\coffee-cache.js`
line 14: `cacheDir = path.join(fs.absolute('~/.atom'), 'compile-cache');`

**Comment by kleerkoat** [comment]
i don't really belong posting here but it's a topic I have been following. cause I to really want to move to Atom.  I'm not a coding/config guru btw.

@nuxonom that worked to get the .atom folder out of roaming, but there is still a plain Atom folder in there. does it contain anything that needs to "travel" to be portable? or is it just a temporary place that will just be recreated on the next computer you plugin into without losing anything? apologies again since this is more of a support question that a contributing one.
![capture](https://cloud.githubusercontent.com/assets/1906101/5966722/cdb745be-a7d4-11e4-94bc-d262094cccb7.PNG)

**Comment by notnsane** [comment]
I don't understand - the available .zip version is the portable version or there are other issues?

**Comment by Ricain** [comment]
Good question, it looks like there is a `atom-windows.zip` on the [release page](https://github.com/atom/atom/releases). But I cannot see anything for Linux...

**Comment by iliyang** [comment]
@Ricain look at #5385. Seems like Atom can already be made "portable" but it requires setting an `ATOM_HOME` environment variable. I'd still like it if Atom would also look for an `.atom` or `.config` or whatever subdirectory in the application directory. This would avoid the hassle of dealing with env vars.

**Comment by Ricain** [comment]
@iliyang thank, I will take a look at it.

**Comment by JixunMoe** [comment]
I wrote my self a simple boot script, and it works fine:

```bat
@echo off
pushd %~dp0
if not exist .atom mkdir .atom
set "ATOM_HOME=%~dp0\.atom"
atom
```

**Comment by iliyang** [comment]
@Ricain Did you have time to investigate the check for a `.atom` (or `.config`) profile directory at launch?

**Comment by spencersager** [comment]
Okay, so now that #5385 is working, how do I set the ATOM_HOME env var? I'm working on a project between school and home off of a flash drive. How can I get this configured? Thanks.

**Comment by proimage** [comment]
In case all you non-Windows people are curious, the "installation" process of portable apps should be nothing but copying/extracting files into a specified folder. No environment variables (which are handled via the OS), registry keys, or entries in the "Add/Remove Programs" listing are allowed.

The use of the user's profile folder is discouraged (i.e. it should not be the default installation location), although not forbidden. Most portable apps I've used default to placing the program's files in a `\[program_name]` folder, which makes it easy to select a thumbdrive as the destination.

All of a program's settings should be stored in a file inside same folder where the program is located, or in a subfolder thereof.

In essence, if you delete the one folder where the program is installed, there should be absolutely no residue or indication that that program was ever on the computer (ignoring delete-recovery aspects such as the Recycle Bin and Windows' self-maintained list of recently run programs, which can't be helped).

**Comment by yuki-takei** [comment]
@JixunMoe 's script wasn't work for me.

environment: Windows 7 + atom-windows.zip(0.209.0)
error message window:
```
A JavaScript error occured in the browser process

Uncaught Exception:
Error: Failed to get path
  at Error (native)
  at Object.<anonymous>
(C:\Users\MyName\somewhere\Atom\resources\atom.asar\browser\lib\init.js:96:41)
  at Object.<anonymous>
(C:\Users\MyName\somewhere\Atom\resources\atom.asar\browser\lib\init.js:104:4)
  at Module._compile (module.js:452:26)
  at Object.Module._extensions..js (module.js:470:10)
  at Module.load (module.js:347:32)
  at Function.Module._load (module.js:302:12)
  at Function.Module.runMain (module.js:493:10)
  at startup (node.js:120:16)
  at node.js:877:3
```

It seems like be caused from that there is no directories `.atom\AppData\Roaming`.

So I changed the script like follows and it works for me.
\# In order to remove cmd window, using `start` :-)

```
@echo off
pushd %~dp0
if not exist .atom mkdir .atom
if not exist .atom\AppData mkdir .atom\AppData
if not exist .atom\AppData\Roaming mkdir .atom\AppData\Roaming

:: Trick emmet or other plugin to think profile directory is in our data directory
set "USERPROFILE=%~dp0\.atom"

:: Set Atom base directory
set "ATOM_HOME=%~dp0\.atom"

:: Boot
start atom
```

**Comment by war59312** [comment]
http://portableapps.com/node/43424

Would love to see a PortableApps version.

**Comment by stavpup** [comment]
"John T. Haller
For future reference, this requires .NET 4.5 which is only bundled with Windows 8 and higher. Windows 7 and lower (the vast majority of Windows machines) requires .NET 4.5 to be installed with administrative rights."

To comment on that, windows 7 machines will upgrade to windows 10, so that is not an issue.
http://portableapps.com/node/43424

**Comment by amroamroamro** [comment]
For non-Windows devs, here are some definitions of what constitutes a portable app:

- http://portableapps.com/about/what_is_a_portable_app
- http://www.portablefreeware.com/about.php
- http://superuser.com/questions/137894/whats-the-difference-between-portable-and-install-versions
- http://www.rlvision.com/blog/what-is-a-portable-application-and-why-should-you-use-them/
- http://www.rarst.net/software/portable-definition/
...

**Comment by docterry** [comment]
This worked for me too! (Also have had success in converting to a PortableApps format, but not ready for primetime). 

The only problem that I have found so far is trying to get linter-php to work. It works with an absolute path to php.exe, such as "\Programs\PortableApps\AtomPortable\App\Atom\php\php.exe". However, when I give a relative path like "App\Atom\php\php.exe" it does not perform the error checking.

**Comment by Ricain** [comment]
@iliyang nope I did not yet and I will probably won't since school stopped: I tried on school's Debian with no root access... (sorry for this late answer)

**Comment by alexchandel** [comment]
**Please reopen.** Atom is not portable. It saves temporary files to the root volume, and the current hacked batch-method isn't deployable.

**Comment by CraigLager** [comment]
This needs re-opening. Installed packages do not transfer from one machine to another.

Tested by installing merge-conflicts package into the zip version extracted into dropbox, then opening on another machine.

**Comment by nelson6e65** [comment]
There is a Linux portable version?

**Comment by bitbonk** [comment]
Why is this closed? It doesn't work for me? I either have to set the ATOM_HOME environment variable on the target machine or else the directory will always ~/Atom . That doesn't make it portable.

**Comment by snuggles08** [comment]
Please reopen.

**Comment by mwyatt** [comment]
Once this becomes portable I can actually start using it, as I switch between two computers each day this is essential for me!

**Comment by nelson6e65** [comment]
@mwyatt, there is a Package to sync configs, packages, etc, between different computers: https://atom.io/packages/sync-settings

It's really useful, easy con configure, using [Gists](https://gist.github.com/), and you can use it for now until Atom Portable is released. :smile:

**Comment by batjko** [comment]
I have my ATOM_HOME variable set to a `Configs/.atom` folder in my Dropbox on all three of my computers, and they are always synced that way.

It's correct that this isn't truly portable, but it's pretty quickly done and does the job.

**Comment by Ricain** [comment]
@batjko :+1: 

I will try with [owncloud](https://owncloud.org/)!

**Comment by mwyatt** [comment]
Thanks @nelson6e65 + @batjko will try these out!

**Comment by impinball** [comment]
:+1: Please reopen this. It would be epicly awesome!

**Comment by Zireael07** [comment]
Please reopen.

**Comment by paulcbetts** [comment]
It seems like we could do some more work to make this easier, maybe detect whether to run in Portable Mode™ based on `process.execPath`

**Comment by Machou** [comment]
need xD

**Comment by batjko** [comment]
@paulcbetts I think what people still wait for is a Portable Mode™ that comes out of the box. For example the zipped release version could be made to be fully self-contained (not relying on env vars etc)

They don't want to have to set any environment variables, use command line switches or whatnot.

**Comment by impinball** [comment]
@batjko That may very well be the best way IMHO.

**Comment by paulcbetts** [comment]
@batjko I think we're violently agreeing here, basically we use `process.execPath` to detect whether we're installed, then if we're not installed, we operate in Portable Mode.

**Comment by proimage** [comment]
The problem with that approach is if Atom is "installed" in portable mode in Dropbox or a USB drive, and an attempt is made to run it on a computer where there is a pre-existing installation of Atom...   you see the issue.

I'd strongly urge you to implement it as a portable app according to the PortableApps.com spec: http://portableapps.com/development

**Comment by batjko** [comment]
Is it a huge effort to maintain a fully independent portable release - i.e. without conditional launch mode checking, and without an installer, just the ready-to-go folder with a local .atom inside the release directory?

**Comment by iliyang** [comment]
I agree that it could be a pain to maintain a PortableApps-style release. I, and I also hope most other users, would be perfectly happy with an `.atom` folder inside the installation folder, as long as everything remains completely encapsulated and no files are written anywhere outside the installation folder. Such a simple structure would also hopefully facilitate automatic updates.

**Comment by paulcbetts** [comment]
> Is it a huge effort to maintain a fully independent portable release - i.e. without conditional launch mode checking, and without an installer, just the ready-to-go folder with a local .atom inside the release directory?

I think it'd probably be super easy, I might try to make a patch for it

**Comment by sketchpunk** [comment]
A year ago I made a copy mostly portable by simply having a .atom folder in the application's folder. The only real issue is duplication and hardcoded values through out the app to determines the location of the .atom folder. Ideally there should be a single global function or variable that holds the location of the .atom folder, then the rest of the app and all plugins get the value from this global func/var. In good api design this is how an app should be setup. I would of done these myself but I dont like coffeescript and I feel it was a mistake to use coffee script in developing atom, but that's another issue entirely. The changes I did was to the javascript of the compiled atom btw. Look at previous comments and examples I made in this thread literally a year ago. Which is kinda sad that after a year this is still an issue.

**Comment by impinball** [comment]
I would prefer to have a simple zip file. It's easier to set up and use
than a PortableApps installer. And I'm already used to a similar workflow
with the Linux global install version. My school uses Windows almost
exclusively, which necessitates having a portable version to take around.
It would be incredibly convenient.

**Comment by proimage** [comment]
No reason not to have both.

FYI, one major advantage PortableApps has over just a zip is that it offers automatic app updates.

**Comment by nengxu** [comment]
+1 for portable mode

**Comment by softwarespot** [comment]
+1

**Comment by kdmaas** [comment]
+1 for portable mode

**Comment by kdmaas** [comment]
@impinball
Am I missing something?
If you don't want to use the PortableApps menu you can use the self-extracting zip-files directly.
You'll have to remember to update manually though.

**Comment by impinball** [comment]
@kdmaas 

It's not so much the menu. I just prefer unzipping a zipball in a few steps through the standard instead of clicking through about 6-10 screens in a pretty ~~useless~~ GUI that does effectively the same thing.

Unless Atom modifies the registry (which it [doesn't](https://github.com/atom/electron/search?utf8=%E2%9C%93&q=registry) [appear to](https://github.com/atom/atom/search?utf8=%E2%9C%93&q=registry), even in a global install), I don't think there would be much advantage to an installer that is just a glorified self-extracting zip file. You might as well use 7-zip for that - it would accomplish the same thing.

I also don't think that there would be a lot of work to make this zip-file portable. IMHO it would be easiest if Atom included a way to change the user path (i.e. the return value of `os.homedir()`) at the time of initialization. It's something that could also be done via setting the `%userprofile%` variable in a way local to the application.

**Comment by JerichoDefeated** [comment]
+1

**Comment by dumpsters** [comment]
+1, only thing keeping me from using atom

**Comment by Sisyphuss** [comment]
+1

**Comment by capSAR273** [comment]
Just chiming in with my support for a portable version, I use Atom off of a flash drive for coursework and losing custom settings/packages is less than optimal. I'll try the suggestions put here by others until Atom is truly portable :+1:

**Comment by jamescostian** [comment]
:+1:

**Comment by Ivan171** [comment]
+1

**Comment by javiglesias** [comment]
+1

**Comment by bhack** [comment]
:+1:

**Comment by thomasjo** [comment]
This is being worked on in https://github.com/atom/atom/pull/8442, so there is no need to +1 this and in so doing, notifying 52 people. As a result, I'm locking this issue for the time being. Feel free to subscribe to the PR if you're curious about the progress being made.

**Comment by paulcbetts** [comment]
This is now merged into master as part of #8442 and friends, and will show up in the next release! Thanks for the feedback all

**Comment by benogle** [comment]
💥 🎉
